# **EJERCICIO 23**


---


Integrantes:
*   Oziel Banda Hernández
*   Joshua Santiago Cruz Pérez
*   Naomi Daniela Jiménez Borzani
*   Ximena Paredes Hernández


---



### **Funciones Auxiliares**

In [ ]:
import numpy as np

def MultMat(Mat1, Mat2):
    """
    Multiplica dos matrices utilizando el método clásico de triple bucle.

    Parámetros:
        Mat1 (np.array): Primera matriz de tamaño m×n
        Mat2 (np.array): Segunda matriz de tamaño n×p

    Retorna:
        np.array: Matriz resultante de tamaño m×p
    """
    Mat3 = np.zeros((Mat1.shape[0], Mat2.shape[1]))
    for row in range(Mat1.shape[0]):
        for col in range(Mat2.shape[1]):
            for aux in range(Mat2.shape[0]):
                Mat3[row, col] += Mat1[row, aux] * Mat2[aux, col]
    return Mat3

def MatVec(Mat, vec):
    """
    Multiplica una matriz por un vector.

    Parámetros:
        Mat (np.array): Matriz de tamaño m×n
        vec (np.array): Vector de tamaño n

    Retorna:
        np.array: Vector resultante de tamaño m
    """
    Mv = np.zeros(Mat.shape[0])
    for ren in range(Mat.shape[0]):
        for col in range(Mat.shape[1]):
            Mv[ren] += Mat[ren, col] * vec[col]
    return Mv

def Transpuesta(Mat):
    """
    Calcula la transpuesta de una matriz cuadrada mediante intercambio in-place.

    Parámetros:
        Mat (np.array): Matriz cuadrada a transponer

    Retorna:
        np.array: Matriz transpuesta (la misma matriz modificada)
    """
    for ren in range(Mat.shape[0]):
        for col in range(Mat.shape[1]):
            if ren < col:
                Mat[ren, col], Mat[col, ren] = Mat[col, ren], Mat[ren, col]
    return Mat

def SubMat(Mat, ren_out, col_out):
    """
    Crea una submatriz eliminando una fila y columna específicas.

    Parámetros:
        Mat (np.array): Matriz original
        ren_out (int): Índice de fila a eliminar
        col_out (int): Índice de columna a eliminar

    Retorna:
        np.array: Submatriz resultante
    """
    return np.delete(np.delete(Mat, ren_out, axis=0), col_out, axis=1)

def Det(Mat):
    """
    Calcula el determinante de forma recursiva usando expansión por menores.

    Parámetros:
        Mat (np.array): Matriz cuadrada

    Retorna:
        float: Valor del determinante

    Notas:
        Casos base para matrices 1×1 y 2×2
        Caso recursivo para matrices n×n con n > 2
    """
    if Mat.shape == (1, 1):
        return Mat[0, 0]
    if Mat.shape == (2, 2):
        return Mat[0, 0]*Mat[1, 1] - Mat[0, 1]*Mat[1, 0]
    deter = 0
    for col in range(Mat.shape[1]):
        menor = SubMat(Mat, 0, col)
        signo = (-1) ** col
        deter += signo * Mat[0, col] * Det(menor)
    return deter

def Cofactores(Mat):
    """
    Calcula la matriz de cofactores.

    Parámetros:
        Mat (np.array): Matriz cuadrada

    Retorna:
        np.array: Matriz de cofactores
    """
    Cofa = np.zeros_like(Mat, dtype=float)
    for ren in range(Mat.shape[0]):
        for col in range(Mat.shape[1]):
            Cofa[ren, col] = ((-1) ** (ren + col)) * Det(SubMat(Mat, ren, col))
    return Cofa

def Inv(Mat):
    """
    Calcula la inversa de una matriz usando el método de cofactores.

    Parámetros:
        Mat (np.array): Matriz cuadrada invertible

    Retorna:
        np.array: Matriz inversa

    Excepciones:
        ValueError: Si la matriz es singular (determinante = 0)
    """
    deter = Det(Mat)
    if deter == 0:
        raise ValueError("La matriz es singular y no tiene inversa.")
    Cofac = Cofactores(Mat)
    Adj = Transpuesta(Cofac.copy())
    return (1 / deter) * Adj

## **Inciso a)**



---

a) Resolver el siguiente problema de mínimos cuadrados usando cualquier método de los vistos en clase.
$$
\left[\begin{array}{ll}
0.16 & 0.10 \\
0.17 & 0.11 \\
2.02 & 1.29
\end{array}\right]\left[\begin{array}{l}
x_1 \\
x_2
\end{array}\right] \approx\left[\begin{array}{l}
0.26 \\
0.28 \\
3.31
\end{array}\right]
$$

---



In [ ]:
"""
Resuelve un sistema de mínimos cuadrados utilizando el método de ecuaciones normales,
con multiplicación e inversión de matrices implementadas manualmente.

Este método encuentra la solución x que minimiza ||Ax - b||² para un sistema sobredeterminado
(más ecuaciones que incógnitas), aplicando la fórmula:

    x = (AᵗA)^(-1) Aᵗb

Pasos:
1. Se define la matriz A y el vector b del sistema original.
2. Se calcula la transpuesta de A.
3. Se multiplica Aᵗ por A.
4. Se multiplica Aᵗ por b.
5. Se calcula la inversa de AᵗA.
6. Se multiplica la inversa de AᵗA por Aᵗb para obtener la solución x.
7. Se imprime la solución obtenida.

Nota:
- Este enfoque puede ser numéricamente inestable si AᵗA está mal condicionada.
"""

# Matriz A y vector b original
A = np.array([
    [0.16, 0.10],
    [0.17, 0.11],
    [2.02, 1.29]
])

b = np.array([
    [0.26],
    [0.28],
    [3.31]
])

# Paso 1: Calcular A^T (transpuesta de A)
A_T = A.T.copy()

# Paso 2: Calcular A^T * A
ATA = MultMat(A_T, A)

# Paso 3: Calcular A^T * b
ATb = MultMat(A_T, b)

# Paso 4: Calcular la inversa de A^T * A
ATA_inv = Inv(ATA)

# Paso 5: Calcular x = (A^T A)^(-1) * A^T b
x = MultMat(ATA_inv, ATb)

# Mostrar el resultado de la solución
print("Inciso a) Solución del sistema original:")
print(x)



Inciso a) Solución del sistema original:
[[1.]
 [1.]]


## **Inciso b)**



---

b) Resolver el mismo problema con la siguiente perturbación en el vector $\bar{b}$,
$$
\bar{b}=\left[\begin{array}{l}
0.27 \\
0.25 \\
3.33
\end{array}\right]
$$

---



In [ ]:
"""
Calcula la solución del problema de mínimos cuadrados para una versión perturbada del vector
b en el sistema Ax ≈ b, utilizando el método de ecuaciones normales.

Este código forma parte del inciso b), ejercicio en el que se analiza el impacto de una
pequeña perturbación en el vector sobre la solución del sistema. Se aprovechan los cálculos
ya realizados en el inciso a), como A^T, (A^T A) y su inversa, para no repetir operaciones.

Requiere previamente:
- Haber definido la matriz A y su transpuesta A_T.
- Haber calculado (A^T A)^(-1), almacenado en ATA_inv.
- Haber definido la función MultMat para multiplicar matrices.
"""

# Vector perturbado
b_pert = np.array([
    [0.27],
    [0.25],
    [3.33]
])

# Paso 1: Calcular A^T * b perturbado
ATb_pert = MultMat(A_T, b_pert)

# Paso 2: Calcular la nueva solución x = (A^T A)^(-1) * A^T * b_perturbado
x_pert = MultMat(ATA_inv, ATb_pert)

# Imprimir resultado
print("Inciso b) Solución con vector perturbado:")
print(x_pert)



Inciso b) Solución con vector perturbado:
[[ 7.00888731]
 [-8.39566299]]


## **Comparación de Soluciones**



---

c) Compara los resultados de los incisos anteriores, ¿se puede explicar la diferencia entre ellos?.

---



Al resolver el sistema de mínimos cuadrados utilizando el método de ecuaciones normales, se obtuvo que una pequeña perturbación en el vector de términos independientes provoca un cambio muy significativo en la solución del sistema. Mientras que la solución original fue:

X = [1.0,1.0]

la solución con la perturbación fue:

X = [7.01, −8.40]

Este comportamiento se puede explicar por el hecho de que el sistema está mal condicionado. Es decir, las columnas de la matriz A son casi linealmente dependientes, lo cual genera una matriz A transpuesta por A cercana a la singularidad, y por tanto, muy sensible a pequeños cambios en los datos.

Este ejemplo nos muestra la importancia de estudiar la estabilidad de los métodos numéricos. Aunque las ecuaciones normales son computacionalmente más simples, no siempre son la mejor opción en términos de estabilidad. Por eso, para problemas reales con datos ruidosos o mal condicionados, se prefieren métodos como descomposición QR (Householder o Givens) que ofrecen mayor estabilidad numérica.


In [ ]:
# Comparación con método de mínimos cuadrados de NumPy

# a) Solución con b original
x_np_a, _, _, _ = np.linalg.lstsq(A, b, rcond=None)
print("Solución con NumPy (original):")
print(x_np_a)

# b) Solución con b perturbado
x_np_b, _, _, _ = np.linalg.lstsq(A, b_pert, rcond=None)
print("Solución con NumPy (perturbado):")
print(x_np_b)


Solución con NumPy (original):
[[1.]
 [1.]]
Solución con NumPy (perturbado):
[[ 7.00888731]
 [-8.39566299]]
